### PR TITLE
expression: use new metadata instead of implicit_args

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

### What problem does this PR solve? 

Currently, implicit_args can only pass data with `[]Datum` format, it's not reasonable for structured metadata, such as for `in` optimization, we need some metadata format like https://github.com/TennyZhuang/tikv/blob/fdb0bb3777f2bb32716a10beb776cc5e1c55125f/components/tidb_query/src/rpn_expr/impl_compare_in.rs#L125

### What is changed and how it works?

This PR introduce arbitrary metadata format defined with protobuf in tipb, encoding in TiDB and decoding in TiKV.

Related to https://github.com/pingcap/tipb/pull/161

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

This feature will break the implicit_args format, but will not break anything really, because only `cast` use implicit_args, and the pushdown switch of `cast` is never open before.

Related changes

 - Need to merge https://github.com/pingcap/tipb/pull/161 first.

Release note

 - Introduce new implicit_args format

### CODE REVIEW

**Comment by TennyZhuang** [comment]
I will remove the `replace` in go.mod after https://github.com/pingcap/tipb/pull/161/ merged

**Comment by TennyZhuang** [comment]
@breeswish

File: expression/expr_to_pb.go [hunk 1]
```diff
@@ -266,20 +267,20 @@ func (pc PbConverter) scalarFuncToPBExpr(expr *ScalarFunction) *tipb.Expr {
...
 
-	var implicitArgs []byte
-	if args := expr.Function.implicitArgs(); len(args) > 0 {
-		encoded, err := codec.EncodeValue(pc.sc, nil, args...)
+	var metadata []byte
+	if args := expr.Function.metadata(); args != nil {
+		encoded, err := proto.Marshal(args)
 		if err != nil {
-			logutil.BgLogger().Error("encode implicit parameters", zap.Any("datums", args), zap.Error(err))
+			logutil.BgLogger().Error("encode implicit parameters", zap.Any("implicit_args", args), zap.Error(err))
```
Comment (breeswish) [inline expression/expr_to_pb.go]: Wrong error message

File: expression/expr_to_pb_test.go [hunk 2]
```diff
@@ -642,28 +641,22 @@ func (s *testEvaluatorSerialSuites) TestImplicitArgs(c *C) {
 
 	// InUnion flag is false in `BuildCastFunction` when `ScalarFuncSig_CastStringAsInt`
 	cast := BuildCastFunction(mock.NewContext(), dg.genColumn(mysql.TypeString, 1), types.NewFieldType(mysql.TypeLonglong))
-	c.Assert(cast.(*ScalarFunction).Function.implicitArgs(), DeepEquals, []types.Datum{types.NewIntDatum(0)})
+	c.Assert(cast.(*ScalarFunction).Function.metadata(), DeepEquals, &tipb.InUnionMetadata{InUnion: false})
 	expr := pc.ExprToPB(cast)
 	c.Assert(expr.Sig, Equals, tipb.ScalarFuncSig_CastStringAsInt)
 	c.Assert(len(expr.Val), Greater, 0)
-	datums, err := codec.Decode(expr.Val, 1)
```
Comment (breeswish) [inline expression/expr_to_pb_test.go]: Assert decoded `expr.Val`?

File: expression/expr_to_pb.go [hunk 3]
```diff
@@ -266,20 +267,20 @@ func (pc PbConverter) scalarFuncToPBExpr(expr *ScalarFunction) *tipb.Expr {
 		children = append(children, pbArg)
 	}
 
-	var implicitArgs []byte
-	if args := expr.Function.implicitArgs(); len(args) > 0 {
-		encoded, err := codec.EncodeValue(pc.sc, nil, args...)
+	var metadata []byte
+	if args := expr.Function.metadata(); args != nil {
```
Comment (breeswish) [inline expression/expr_to_pb.go]: name it metadata?

Comment (TennyZhuang) [reply expression/expr_to_pb.go]: No variable shadow in golang, metadata already exists in L270..

Comment (breeswish) [reply expression/expr_to_pb.go]: encoded_metadata XD

Comment (TennyZhuang) [reply expression/expr_to_pb.go]: encodedMetadata

Comment (TennyZhuang) [reply expression/expr_to_pb.go]: use encoded

**Comment by breeswish** [comment]
LGTM

**Comment by TennyZhuang** [comment]
@wshwsh12 @XuHuaiyu PTAL

**Pull request merged**